# Module 59.1: Yield-Based Bond Duration Measures and Properties

<hr>

In [44]:
# Import libraries
import math
import pandas as pd
import numpy as np
import numpy_financial as npf
# import matplotlib.pyplot as plt
# import seaborn as sns
# from sklearn.model_selection import train_test_split
# from pandas.core.common import random_state
# from sklearn.linear_model import LinearRegression

### LOS 59.a: Define, calculate, and interpret modified duration, money duration, and the price value of a basis point (PVBP).

Modified duration (ModDur) is calculated as Macaulay duration (MacDur) divided by one plus the bond's periodic rate of return YTM divided by periodicity).

For an annual-pay bond, this is the general form of ModDur:

$$
\text{ModDur} = \frac{\text{MacDur}}{(1 + YTM)}
$$

.

For a semiannual-pay bond with a YTM quoted on a semiannual bond basis, this is the form:

$$
\text{ModDurSEMI} = \frac{\text{MacDur_SEMI}}{(1 + YTM)\div 2}
$$


**EXAMPLE: MODIFIED DURATION**

A 5-year, 11% annual coupon bond priced at 86.59 to yield 15% to maturity has a Macaulay duration of 4.03. Calculate the modified duration of this bond.

In [10]:
# Calculate the modified duration of this bond.

Mac_Dur = 4.03
ModDur = Mac_Dur / (1 + 0.15)
print(round(ModDur,2))


# Print PV of 0.5% increase in YTM (part 1)
PV_ModDur1 = round(ModDur,2) * 0.5
print(PV_ModDur1)

# Print PV of 0.5% increase in YTM (part 2)
PV_ModDur2 = round(npf.pv(nper=5, fv=100, pmt=11, rate=.155),3)
print(PV_ModDur2)

3.5
1.75
-85.092


**Answer:**

Because it is an annual coupon bond (periodicity = 1), its modified duration can be calculated as follows:

`ModDur = 4.03 / 1.15 = 3.50`

Modified duration provides an estimate for the percentage change in a bond's price given a 1\% change in YTM:

$\qquad\text{approximate percentage change in bond price} = \text{–ModDur} × \Delta \text{YTM}$

Based on a `ModDur` of `3.50`, in response to an 0.5\% increase in YTM the price of the bond should fall by approximately `3.50 × 0.5% = 1.75%`. The resulting price estimate of `86.59 × (1 – 0.0175) = 85.075` is close to the value of the bond calculated directly using a YTM of `15.5%`, which is `85.092`:

`N = 5; I/Y = 15.5; FV = 100; PMT = 11; CPT → PV = –85.092`

For a semiannual-pay bond, ModDurSEMI can be annualized (from semiannual periods to annual periods) by dividing by two, and then used as the approximate change in price for a 1% change in a bond's YTM.

#### Approximate Modified Duration

We can approximate modified duration directly using bond values for an increase in YTM and for a decrease in YTM of the same size.

The calculation of approximate ModDur is based on a given change in YTM. V– is the price of the bond if YTM is decreased by ΔYTM, and V+ is the price of the bond if the YTM is increased by ΔYTM. Approximate ModDur is given by the following:

\
$\text{approximate ModDur} = \Large\frac{V_{-} - V_{+}}{2 \times \text{V} × \Delta \text{YTM}}$

Where:

$V_{-}$ = Bond price at lower yield \
$V_{+}$ = Bond price at 1% higher yield \
$V$ = Bond price at current yield \
$\Delta \text{YTM}$ = Change in yield to maturity


The formula uses the average of the magnitudes of the price increase and the price decrease, which is why $V_{–} − V_{+}$ in the numerator is divided by two in the denominator.

$V_{0}$, the current price of the bond, is in the denominator to convert this average price change to a percentage, and the  $\Delta \text{YTM}$ term is in the denominator to scale the duration measure to a 1\% change in yield by convention. Note that the $\Delta \text{YTM}$ term in the denominator must be entered as a decimal (rather than in a whole percentage) to properly scale the duration estimate

**EXAMPLE: CALCULATING APPROXIMATE MODIFIED DURATION**

Consider a 5-year, 11% annual coupon bond priced at 86.59 to yield 15\% to maturity. If its YTM increases by 50 basis points, its price will decrease to 85.092. If its YTM decreases by 50 basis points, its price will increase to 88.127. Calculate the approximate `ModDur3`.

In [31]:
# EXAMPLE: CALCULATING APPROXIMATE MODIFIED DURATION
Vminus =  88.127 
Vplus =  85.092
V = 86.59
YTM = 0.005

# ModDur3 = (88.127 - 85.092) / (2 * 86.59 * 0.005)
ModDur3 = (Vminus - Vplus) / (2 * V * YTM)

print(round(ModDur3,4))

3.505


Modified duration is a *linear estimate* of the relation between a bond's price and YTM, whereas the actual relation is convex—not a straight line. This means that the modified duration measure provides good estimates of bond prices for small changes in yield, but increasingly poor estimates for larger changes in yield as the effect of the curvature of the price-yield curve is more pronounced. We illustrate this in  **Approximate ModDur**.

*Approximate ModDur*

<img src="https://github.com/PachaTech/CFA-Level-1/blob/main/Fixed%20Income/Module%2059/pics/duration_est.jpeg?raw=true"></img>

#### Money Duration

The money duration of a bond position (also called dollar duration) is expressed in currency units:


$$
\text{money duration} = \text{annual ModDur} × \text{full price of bond position}
$$


Multiplying the money duration of a bond by a given change in YTM (as a decimal) will provide an estimate for the change in bond value for that change in YTM.

<hr>

**EXAMPLE: MONEY DURATION**

Calculate the money duration on a coupon date of a \\$ 2 million par value bond that has a ModDur of 7.42 and a full price of 101.32, expressed for the whole bond and per \\$100 of face value.
What will be the impact on the value of the bond of a 25 basis point increase in its YTM?

In [42]:
# EXAMPLE: MONEY DURATION
MoneyDur = 7.42 * 2000000 * 1.0132
print(round(MoneyDur))

# The money duration per $100 of par value is:
print(7.42 * 101.32)

# or Alternatively
print(15035888 / (2000000 / 100))

print(15035888 * 0.0025)
print("The bond value decreases by $",(15035888 * 0.0025))

15035888
751.7944
751.7944
37589.72
The bond value decreases by $ 37589.72


**Answer:**

The money duration for the bond is ModDur times the full value of the bond:
$7.42 × \$2,000,000 × 101.32\% = \$15,035,888$

The money duration per $100 of par value is:

$7.42 × 101.32 = \$751.79$

Or, 

$\large\frac{\$15,035,888}{(\$2,000,000 \div \$100)} \small = \$751.79.$

$\$15,035,888 × 0.0025 = \$37,589.72$

The bond value decreases by $\$37,589.72$

<br>

The **price value of a basis point (PVBP)** is the money change in the full price of a bond when its YTM changes by one basis point, or 0.01\%. We can calculate the PVBP directly for a bond by calculating the average of the decrease in the full value of a bond when its YTM increases by one basis point, and the increase in the full value of the bond when its YTM decreases by one basis point.

<br/>

<hr>

**EXAMPLE: CALCULATING THE PVBP**

A newly issued, 20-year, 6\% annual-pay straight bond is priced at 101.39. Calculate the `PVBP` for this bond, assuming it has a par value of \$1 million.


In [75]:
# EXAMPLE: CALCULATING THE PVBP

# First, we need to find the YTM of the bond:
bond_YTM = npf.rate(nper=20, pmt=6, pv= -101.39,fv=100) * 100
print(round(bond_YTM,2))


# Now, we need the values for the bond with YTMs of 5.89 and 5.87:
V_plus = npf.pv(nper= 20,pmt=6, rate= 0.0589,fv=100) *-1
print(round(V_plus,3))

# I/Y = 5.87; CPT → PV = –101.507 (V–)
V_minus = npf.pv(nper= 20,pmt=6, rate= 0.0587,fv=100) *-1
print(round(V_minus,3))

# PVBP (per $100 of par value) = (101.507 – 101.273) / 2 = 0.117
PVBP = (V_minus - V_plus) / 2
print(round(PVBP,3))

print(round(1000000 * PVBP * 0.01))

5.88
101.273
101.507
0.117
1170


**Answer:**

First, we need to find the YTM of the bond:

    N = 20; PV = –101.39; PMT = 6; FV = 100; CPT→ I/Y = 5.88

Now, we need the values for the bond with YTMs of 5.89 and 5.87:

    I/Y = 5.89; CPT → PV = –101.273 (V+)

    I/Y = 5.87; CPT → PV = –101.507 (V–)

PVBP (per \\$100 of par value) = (101.507 – 101.273) / 2 = 0.117

For the \\$1 million par value bond, each 1 basis point change in the YTM will change the bond's price by 0.117 × \\$1 million × 0.01 = \\$1,170.



### LOS 59.b: Explain how a bond's maturity, coupon, and yield level affect its interest rate risk.

Other things equal, bonds with *longer maturity* will (usually) have higher interest rate risk. The present values of payments made further in the future are more sensitive to changes in the discount rate used to calculate present value than are the present values of payments made sooner.

We must say usually because there are instances where an increase in a discount coupon bond's maturity will decrease its Macaulay duration. For a discount bond, duration first increases with longer maturity and then decreases over a range of relatively long maturities until it approaches the duration of a perpetuity, which is `(1 + YTM) / YTM`.

Between coupon dates, if the YTM of a coupon bond remains constant, its Macaulay duration decreases smoothly with the passage of time, and then goes back up slightly at each coupon payment date as the time to the next coupon resets to a full coupon period.

Other things equal, a *higher coupon rate* on a bond will decrease its interest rate risk. For a given maturity and YTM, the duration of a zero-coupon bond will be greater than that of a coupon bond. Increasing the coupon rate means more of a bond's value will be from payments received sooner, so that the value of the bond will be less sensitive to changes in yield.

For floating-rate notes (FRNs) where coupons are periodically reset to a market reference rate (MRR), when interest rates rise, the coupon will also rise, limiting the price risk of the bond. Macaulay duration for an FRN can be calculated as the time to the next coupon reset date at the end of the current coupon period.

Other things equal, an *increase (decrease)* in a bond's YTM will decrease (increase) its interest rate risk. To understand this, we can look to the convexity of the price-yield curve and use its slope as our proxy for interest rate risk. At lower yields, the price-yield curve has a steeper slope, indicating that price is more sensitive to a given change in yield.

## Reading 59: Key Concepts

<hr>

### LOS 59.a

Modified duration is a linear estimate of the percentage change in a bond's price that would result from a 1% change in its YTM:

    ModDur = MacDur / (1 + periodic return of bond)

For an expected change in yield, ΔYTM, the expected change in the bond's price is given by the following:

approximate percentage change in bond price = –ModDur × ΔYTM

Modified duration can be approximated by repricing the bond at different yields:

$\text{approximate ModDur} = \Large\frac{V_{-} - V_{+}}{2 \times \text{V} × \Delta \text{YTM}}$


Money duration is stated in currency units and is sometimes expressed per 100 of bond value:

    money duration = annual ModDur × full price of bond position

The price value of a basis point is the change in the value of a bond, expressed in currency units, for a change in YTM of one basis point:

    PVBP = [(V– − V+) / 2] × par value × 0.01

PVBP can also be calculated as money duration × 0.0001.

### LOS 59.b

Holding other factors constant, the following are true:

* Duration increases when maturity increases.
* Duration decreases when the coupon rate increases.
* Duration decreases when YTM increases.
* Duration decreases as time passes, but increases slightly on coupon dates.

## Qbank 59.1 Quiz

<hr>

**Question 1 of 4:** 

A 14% annual-pay coupon bond has six years to maturity. The bond is currently trading at par. Using a 25 basis point change in yield, the approximate modified duration of the bond is closest to:

**ANSWER:** \
3.888

**EXPLANATION:**

$V_{–}$ = 100.979

    N = 6; PMT = 14.00; FV = 100; I/Y = 13.75; CPT → PV = –100.979

$V_{+}$ = 99.035

$I/Y$ = 14.25; CPT → PV = –99.035

$V_0$ = 100.000

$\Delta y$ = 0.0025

So, approximate 

$$
\text{approximate ModDur} = \frac{V_{-} - V_{+}}{2 \times \text{V} × \Delta \text{YTM}} = \frac{100.979−99.035)}{2(100)(0.0025)}=3.888.
$$


\
(Module 59.1, LOS 59.a)


In [76]:
# Question 1 of 4:

ModDur4 = (100.979 - 99.035) / (2 * 100 * 0.0025)

print(round(ModDur4,4))

3.888


**Question 2 of 4:** 

The current price of a \\$1,000, 7-year, 5.5\% semiannual coupon bond is \\$1,029.23. The bond's price value of a basis point is closest to:

In [132]:
# Question 2 of 4
# EXAMPLE: CALCULATING THE PVBP

# First, we need to find the YTM of the bond:
bond_YTM1 = npf.rate(nper=14, pmt=27.5, pv= -1029.23,fv=1000) * 100
print(round(bond_YTM1,5))

# V_plus1 is the PV -1029.23 -> I/Y = 2.49998 x 2 = 4.99995.
V_plus_rate = npf.rate(nper= 14,pmt=27.5, pv= -1029.23 ,fv=1000)*2
print(round(V_plus_rate,7))

# V_minusPV is rounded V_Plus to 5% + 1 bps divided by 2 = 2.505
V_plusPV = npf.pv(nper= 14,pmt=27.5, rate= 0.02505,fv=1000) *-1
print(round(V_plusPV,2))

# PVBP (per $1000 of par) = (1029.23 – 1028.63)
PVBP2 = 1029.23 - 1028.63 
print(round(PVBP2,2))

2.49998
0.0499995
1028.63
0.6


**ANSWER:** \
+8.657\%.

**EXPLANATION:** \
\
`PVBP = initial price – price if yield is changed by 1 basis point.`

First, we need to calculate the yield so we can calculate the price of the bond with a 1 basis point change in yield. Using a financial calculator, `PV = –1,029.23; FV = 1,000; PMT = 27.5 = (0.055 × 1,000) / 2; N = 14 = 2 × 7 years`; and `CPT → I/Y = 2.49998`, multiplied by 2 = 4.99995, or 5.00\%.

Next, compute the price of the bond at a yield of 5.00\% + 0.01\%, or 5.01\%. Using the calculator: `FV = 1,000; PMT = 27.5; N = 14; I/Y = 2.505 (5.01 / 2); CPT → PV = $1,028.63`.

Finally, 
    `PVBP = $1,029.23 – $1,028.63 = $0.60`.

(Module 59.1, LOS 59.a)

**Question 3 of 4:** 

The modified duration of a bond is 7.87. The approximate percentage change in price using duration only for a yield decrease of 110 basis points is closest to:

In [136]:
# Question 3 of 4:

Q3 = -7.87 * (-0.011) 
print(Q3 * 100)

8.657


**ANSWER:** \
+8.657\%.

**EXPLANATION:** \
–7.87 × (–1.10%) = 8.657\%
\
(Module 59.1, LOS 59.a)

**Question 4 of 4:** 

All else equal, which of the following bonds is likely to have the highest price risk?

**ANSWER:**
\
A 2-year zero-coupon bond.

**EXPLANATION:** \
The price risk of the FRN is very low because at the next coupon payment date, the coupons will reset to market rates, and the FRN price will reset to par. Lower coupons, all else equal, lead to greater price risk. Therefore the 2-year zero-coupon bond will have more price risk than the 2-year 10% semiannual-pay bond.

(Module 59.1, LOS 59.b)